File: loan-approval-backend/src/main/java/com/loan/approval/service/ValidationEngine.java
package com.loan.approval.service;

import com.loan.approval.dto.enums.RiskLevel;
import com.loan.approval.dto.response.ExtractedData;
import com.loan.approval.dto.response.ValidationResult;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

@Service
@Slf4j
public class ValidationEngine {

    private static final Pattern PAN_PATTERN = Pattern.compile("[A-Z]{5}[0-9]{4}[A-Z]");
    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$");
    private static final int MIN_AGE = 18;
    private static final int MAX_AGE = 60;
    private static final double MIN_SALARY = 25000;
    private static final int LOAN_TO_SALARY_MULTIPLIER = 60;
    private static final List<String> VALID_OCCUPATIONS = List.of(
            "engineer", "doctor", "teacher", "software engineer", "accountant",
            "manager", "analyst", "consultant", "business owner", "government employee",
            "banker", "lawyer", "architect", "nurse", "scientist"
    );

    public ValidationResult validate(ExtractedData data) {
        List<String> errors = new ArrayList<>();

        validateField(data.getName(), "name", "Name is required", errors);
        validateField(data.getAge(), "age", "Age is required", errors);
        validateField(data.getSalary(), "salary", "Salary is required", errors);
        validateField(data.getOccupation(), "occupation", "Occupation is required", errors);
        validateField(data.getLoanAmount(), "loanAmount", "Loan amount is required", errors);
        validateField(data.getPan(), "pan", "PAN is required", errors);
        validateField(data.getEmail(), "email", "Email is required", errors);
        validateField(data.getAddress(), "address", "Address is required", errors);
        validateField(data.getEmployer(), "employer", "Employer is required", errors);

        if (data.getAge() != null && (data.getAge() < MIN_AGE || data.getAge() > MAX_AGE)) {
            errors.add("Age must be between " + MIN_AGE + " and " + MAX_AGE);
        }

        if (data.getSalary() != null && data.getSalary() <= MIN_SALARY) {
            errors.add("Salary must be greater than " + MIN_SALARY);
        }

        if (data.getPan() != null && !PAN_PATTERN.matcher(data.getPan().toUpperCase()).matches()) {
            errors.add("PAN must be in format AAAAA9999A (5 letters, 4 digits, 1 letter)");
        }

        if (data.getLoanAmount() != null && data.getSalary() != null) {
            double maxLoan = data.getSalary() * LOAN_TO_SALARY_MULTIPLIER;
            if (data.getLoanAmount() > maxLoan) {
                errors.add("Loan amount (₹" + String.format("%.0f", data.getLoanAmount()) +
                        ") exceeds maximum of " + LOAN_TO_SALARY_MULTIPLIER +
                        "x salary (₹" + String.format("%.0f", maxLoan) + ")");
            }
        }

        if (data.getEmail() != null && !EMAIL_PATTERN.matcher(data.getEmail()).matches()) {
            errors.add("Email format is invalid");
        }

        if (data.getOccupation() != null) {
            boolean occupationValid = VALID_OCCUPATIONS.stream()
                    .anyMatch(occ -> data.getOccupation().toLowerCase().contains(occ));
            if (!occupationValid) {
                errors.add("Occupation '" + data.getOccupation() + "' is not recognized as a valid occupation");
            }
        }

        int riskScore = calculateRiskScore(data, errors);
        RiskLevel riskLevel = calculateRiskLevel(riskScore, errors);

        return ValidationResult.builder()
                .errors(errors)
                .riskScore(riskScore)
                .riskLevel(riskLevel)
                .valid(errors.isEmpty())
                .build();
    }

    private void validateField(Object value, String fieldName, String message, List<String> errors) {
        if (value == null || (value instanceof String && ((String) value).isBlank())) {
            errors.add(fieldName + ": " + message);
        }
    }

    private int calculateRiskScore(ExtractedData data, List<String> errors) {
        int score = 0;

        if (errors.isEmpty()) score += 20;

        if (data.getAge() != null) {
            if (data.getAge() >= 25 && data.getAge() <= 45) score += 15;
            else if (data.getAge() > 45 && data.getAge() <= 55) score += 10;
            else score += 5;
        }

        if (data.getSalary() != null) {
            if (data.getSalary() >= 100000) score += 25;
            else if (data.getSalary() >= 50000) score += 20;
            else score += 10;
        }

        if (data.getLoanAmount() != null && data.getSalary() != null) {
            double ratio = data.getLoanAmount() / (data.getSalary() * 12);
            if (ratio <= 2) score += 20;
            else if (ratio <= 4) score += 15;
            else score += 5;
        }

        if (data.getOccupation() != null) {
            String occ = data.getOccupation().toLowerCase();
            if (occ.contains("engineer") || occ.contains("doctor") || occ.contains("scientist")) score += 10;
            else if (occ.contains("manager") || occ.contains("analyst") || occ.contains("consultant")) score += 8;
            else score += 5;
        }

        if (data.getEmployer() != null) score += 5;
        if (data.getAddress() != null) score += 5;

        return Math.min(score, 100);
    }

    private RiskLevel calculateRiskLevel(int score, List<String> errors) {
        if (!errors.isEmpty()) return RiskLevel.HIGH;
        if (score >= 70) return RiskLevel.LOW;
        if (score >= 45) return RiskLevel.MEDIUM;
        return RiskLevel.HIGH;
    }
}

File: loan-approval-backend-node/src/services/validationEngine.ts
import { backendLogger } from '../utils/logger';

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  riskScore: number;
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH';
}

const VALID_OCCUPATIONS = [
  'engineer', 'doctor', 'teacher', 'software engineer', 'accountant',
  'manager', 'analyst', 'consultant', 'business owner', 'government employee',
  'banker', 'lawyer', 'architect', 'nurse', 'scientist', 'businessman',
];

const PAN_REGEX = /^[A-Z]{5}[0-9]{4}[A-Z]$/;
const EMAIL_REGEX = /^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;
const PHONE_REGEX = /^[6-9]\d{9}$/;
const AADHAAR_REGEX = /^\d{12}$/;

export function validateApplication(data: Record<string, any>): ValidationResult {
  const errors: string[] = [];

  if (!data.applicantName || String(data.applicantName).trim() === '') {
    errors.push('Applicant name is required');
  }

  if (!data.pan || String(data.pan).trim() === '') {
    errors.push('PAN is required');
  } else if (!PAN_REGEX.test(String(data.pan).toUpperCase())) {
    errors.push('PAN must be in format AAAAA9999A (5 letters, 4 digits, 1 letter)');
  }

  if (data.aadhaar && !AADHAAR_REGEX.test(String(data.aadhaar))) {
    errors.push('Aadhaar must be a 12-digit number');
  }

  if (data.phone && !PHONE_REGEX.test(String(data.phone))) {
    errors.push('Phone must be a valid 10-digit Indian mobile number');
  }

  if (!data.email || String(data.email).trim() === '') {
    errors.push('Email is required');
  } else if (!EMAIL_REGEX.test(String(data.email))) {
    errors.push('Email format is invalid');
  }

  if (!data.address || String(data.address).trim() === '') {
    errors.push('Address is required');
  }

  if (!data.occupation || String(data.occupation).trim() === '') {
    errors.push('Occupation is required');
  } else if (!VALID_OCCUPATIONS.some(o => String(data.occupation).toLowerCase().includes(o))) {
    errors.push('Occupation not recognized as a valid occupation');
  }

  if (!data.monthlyIncome || Number(data.monthlyIncome) <= 0) {
    errors.push('Monthly income is required and must be greater than 0');
  } else if (Number(data.monthlyIncome) < 25000) {
    errors.push('Monthly income must be at least ₹25,000');
  }

  if (!data.loanAmount || Number(data.loanAmount) <= 0) {
    errors.push('Loan amount is required and must be greater than 0');
  } else if (Number(data.loanAmount) > Number(data.monthlyIncome) * 60) {
    errors.push('Loan amount exceeds 60 times monthly income');
  }

  if (!data.employer || String(data.employer).trim() === '') {
    errors.push('Employer is required');
  }

  let riskScore = 50;
  if (errors.length === 0) riskScore += 10;

  if (data.monthlyIncome >= 100000) riskScore += 15;
  else if (data.monthlyIncome >= 50000) riskScore += 10;
  else riskScore += 5;

  if (data.loanAmount && data.monthlyIncome) {
    const ratio = data.loanAmount / (data.monthlyIncome * 12);
    if (ratio <= 2) riskScore += 15;
    else if (ratio <= 4) riskScore += 10;
    else riskScore += 5;
  }

  if (data.occupation) {
    const occ = String(data.occupation).toLowerCase();
    if (occ.includes('engineer') || occ.includes('doctor') || occ.includes('scientist')) riskScore += 10;
    else if (occ.includes('manager') || occ.includes('analyst')) riskScore += 8;
    else riskScore += 5;
  }

  riskScore = Math.min(riskScore, 100);

  let riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' = 'HIGH';
  if (errors.length === 0 && riskScore >= 70) riskLevel = 'LOW';
  else if (errors.length === 0 && riskScore >= 45) riskLevel = 'MEDIUM';

  const result: ValidationResult = { valid: errors.length === 0, errors, riskScore, riskLevel };
  if (errors.length > 0) {
    backendLogger.warn(`Validation failed: ${errors.length} error(s)`, {
      file: 'src/services/validationEngine.ts',
      function: 'validateApplication',
      details: errors.join('; '),
    });
  } else {
    backendLogger.info(`Validation passed, risk: ${riskLevel} (score: ${riskScore})`, {
      file: 'src/services/validationEngine.ts',
      function: 'validateApplication',
    });
  }
  return result;
}
